File: app/src/main/java/com/banew/cw2025_client/data/DataSource.java
package com.banew.cw2025_client.data;

import static android.content.Context.MODE_PRIVATE;
import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.banew.cw2025_backend_common.dto.coursePlans.CoursePlanBasicDto;
import com.banew.cw2025_backend_common.dto.users.UserLoginForm;
import com.banew.cw2025_backend_common.dto.users.UserProfileBasicDto;
import com.banew.cw2025_backend_common.dto.users.UserTokenFormResult;
import com.banew.cw2025_client.ui.greetings.GreetingsActivity;
import com.banew.cw2025_client.R;
import com.banew.cw2025_client.data.api.ApiService;
import com.google.gson.annotations.SerializedName;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;
import retrofit2.http.Header;

public class DataSource {
    private final SharedPreferences prefs;
    //private final AppDatabase db;

    private static Retrofit retrofit = null;
    private static final String BASE_URL = "http://10.0.2.2:8080/api/";
    private static final boolean isNgrok = false;
    private final Context context;

    private static Retrofit getClient() {
        if (retrofit == null) {
            retrofit = buildClient(BASE_URL);
        }
        return retrofit;
    }

    private static Retrofit buildClient(String path) {
        OkHttpClient client = new OkHttpClient.Builder()
                .connectTimeout(5, TimeUnit.SECONDS)
                .readTimeout(3, TimeUnit.SECONDS)
                .build();

        retrofit = new Retrofit.Builder()
                .baseUrl(path)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        return retrofit;
    }

    private static ApiService getApiService() {
        return getClient().create(ApiService.class);
    }

    public DataSource(Context applicationContext) {
        context = applicationContext;

        prefs = applicationContext.getSharedPreferences("MyAppPrefs", MODE_PRIVATE);

        if (isNgrok) {
            NgrokTokenExtractor.extractNgrokPath("", path -> {
                retrofit = buildClient(path + "/api/");
            });
        }
    }

    public CompletableFuture<Result<List<CoursePlanBasicDto>>> getCurrentCoursePlanList() {
        CompletableFuture<Result<List<CoursePlanBasicDto>>> result = new CompletableFuture<>();

        enqueue(
                getApiService().currentCoursePlanList("Bearer " + getToken()),
        list -> {
                    result.complete(new Result.Success<>(list));
                },
                (t) -> {
                    result.complete(new Result.Error<>(new IOException(
                            context.getString(R.string.network_error), t)));
                },
                Map.of(403, this::logout)
        );

        return result;
    }

    public CompletableFuture<Result<UserProfileBasicDto>> getCurrentUserProfile() {
        CompletableFuture<Result<UserProfileBasicDto>> result = new CompletableFuture<>();

        if (getToken() == null) {
            logout();
        }

        enqueue(
                getApiService().currentUser("Bearer " + getToken()),
                user -> {
                    result.complete(new Result.Success<>(user));
                },
                (t) -> {
                    result.complete(new Result.Error<>(new IOException(
                            context.getString(R.string.network_error), t)));
                },
                Map.of(403, this::logout)
        );

        return result;
    }

    public CompletableFuture<Result<UserTokenFormResult>> login(String username, String password) {
        CompletableFuture<Result<UserTokenFormResult>> result = new CompletableFuture<>();

        var form = new UserLoginForm(username, password);
        enqueue(
                getApiService().login(form),
                resBody -> {
                    result.complete(new Result.Success<>(resBody));
                    updateToken(resBody.token());
                },
                (t) -> {
                    result.complete(new Result.Error<>(new IOException(
                            context.getString(R.string.network_error), t)));
                },
                Map.of(400, () -> {
                    result.complete(new Result.Error<>(new IOException(
                            context.getString(R.string.login_error))));
                })
        );

        return result;
    }

    private <T> void enqueue(Call<T> call,
                             Consumer<T> onSuccessBody,
                             Consumer<Throwable> onInternetError,
                             Map<Integer, Runnable> onFailure) {
        call.enqueue(new Callback<T>() {
            @Override
            public void onResponse(
                    @NonNull Call<T> call,
                    @NonNull Response<T> response
            ) {
                if (response.isSuccessful() && response.body() != null) {
                    onSuccessBody.accept(response.body());
                }
                else {
                    Runnable r = onFailure.get(response.code());
                    if (r != null) r.run();
                }
            }
            @Override
            public void onFailure(@NonNull Call<T> call, @NonNull Throwable t) {
                onInternetError.accept(t);
            }
        });
    }

    @Nullable
    public String getToken() {
        return prefs.getString("jwt_token", null);
    }

    private void updateToken(@Nullable String token) {
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString("jwt_token", token);
        editor.apply();
    }

    public void logout() {
        // TODO
        // так робити, казали, нізя
        Intent intent = new Intent(context, GreetingsActivity.class);
        intent.setFlags(FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
    }

    private static class NgrokTokenExtractor {
        public static void extractNgrokPath(String token, Consumer<String> consumer) {
            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl("https://api.ngrok.com/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();

            NgrokApiService service = retrofit.create(NgrokApiService.class);

            try {
                service
                .getTunnels("Bearer " + token, "2")
                .enqueue(new Callback<NgrokTunnelsResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<NgrokTunnelsResponse> call,
                                           @NonNull Response<NgrokTunnelsResponse> response) {
                        if (response.isSuccessful() && response.body() != null) {
                            String path = response.body().tunnels.stream()
                                    .findFirst()
                                    .orElseThrow()
                                    .publicUrl;

                            System.out.println("Я ЖИВИЙЙЙ\n\n\n\n" + path);
                            consumer.accept(path);
                        }
                    }
                    @Override
                    public void onFailure(@NonNull Call<NgrokTunnelsResponse> call,
                                          @NonNull Throwable t) {

                    }
                });
            }
            catch (Exception e) {
                throw new RuntimeException("Oh shit!", e);
            }
        }

        private interface NgrokApiService {
            @GET("tunnels")
            Call<NgrokTunnelsResponse> getTunnels(
                    @Header("Authorization") String bearerToken,
                    @Header("Ngrok-Version") String version
            );
        }

        private static class NgrokTunnelsResponse {
            public List<NgrokTunnel> tunnels;
        }

        private static class NgrokTunnel {
            @SerializedName("public_url")
            public String publicUrl;
        }
    }
}

File: app/src/main/java/com/banew/cw2025_client/data/DataSource.kt
package com.banew.cw2025_client.data

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.core.content.edit
import com.banew.cw2025_backend_common.dto.BasicResult
import com.banew.cw2025_backend_common.dto.FieldExceptionResult
import com.banew.cw2025_backend_common.dto.cards.FlashCardAnswer
import com.banew.cw2025_backend_common.dto.cards.FlashCardBasicDto
import com.banew.cw2025_backend_common.dto.cards.FlashCardDayStats
import com.banew.cw2025_backend_common.dto.coursePlans.CoursePlanBasicDto
import com.banew.cw2025_backend_common.dto.courses.CourseBasicDto
import com.banew.cw2025_backend_common.dto.courses.CourseDetailedDto
import com.banew.cw2025_backend_common.dto.courses.TopicCompendiumDto
import com.banew.cw2025_backend_common.dto.users.UserLoginForm
import com.banew.cw2025_backend_common.dto.users.UserProfileDetailedDto
import com.banew.cw2025_backend_common.dto.users.UserRegisterForm
import com.banew.cw2025_backend_common.dto.users.UserTokenFormResult
import com.banew.cw2025_client.data.api.ApiService
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import okhttp3.OkHttpClient
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.lang.reflect.Type
import java.time.Duration
import java.time.Instant
import java.util.concurrent.TimeUnit

class DataSource(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("MyAppPrefs", Context.MODE_PRIVATE)

    private suspend fun <T> resolveResult(successCallback: (T) -> Unit = {}, resultSource: (suspend () -> T)) = try {
        val result = resultSource()
        successCallback(result)
        Result.Success(result)
    } catch (e: HttpException) {
        Result.Error(IOException(resolveHttpException(e)))
    } catch (e: Exception) {
        Result.Error(IOException("Network error", e))
    }

    suspend fun beginCourse(coursePlanId: Long): Result<CourseBasicDto> {
        return resolveResult {
            apiService.beginCourse("Bearer $token", coursePlanId)
        }
    }

    suspend fun endCourse(coursePlanId: Long): Result<CourseDetailedDto> {
        return resolveResult {
            apiService.endCourse("Bearer $token", coursePlanId)
        }
    }

    suspend fun createCoursePlan(
        name: String,
        desc: String,
        topics: List<TopicForm>
    ): Result<CoursePlanBasicDto> {
        return resolveResult {
            apiService.createCoursePlan("Bearer $token", CoursePlanBasicDto(
                null,
                name,
                null,
                desc,
                topics.map { CoursePlanBasicDto.TopicBasicDto(
                    null, it.name.value, it.desc.value
                ) }
            ))
        }
    }

    private fun resolveHttpException(e: HttpException) : String =
        when (e.code()) {
            401 -> {
                if (e.code() == 401) logout()
                "HTTP ${e.code()}"
            }
            400 -> {
                val stringObj = e.response()?.errorBody()?.string()
                val message =
                    if (stringObj != null) {
                        try {
                            val obj = Gson()
                                .fromJson(stringObj, FieldExceptionResult::class.java)

                            "${obj.message()}\n${obj.fieldErrors.joinToString(separator = "\n")
                            { "${it.field} - ${it.message}" }}"
                        }
                        catch (_: NullPointerException) {
                            Gson()
                                .fromJson(stringObj, BasicResult::class.java)
                                .message()
                        }
                    }
                    else "Введені дані некоректні!"

                message
            }
            else -> "HTTP ${e.code()}"
        }

    suspend fun currentCourseList(): Result<List<CourseBasicDto>> {
        return resolveResult {
            apiService.getUserCourses("Bearer $token")
        }
    }

    suspend fun courseDetailedById(coursePlanId: Long): Result<CourseDetailedDto> {
        return resolveResult {
            apiService.getCourse("Bearer $token", coursePlanId)
        }
    }

    suspend fun currentCoursePlanList(query: String): Result<List<CoursePlanBasicDto>> {
        return resolveResult {
            apiService.currentCoursePlanList("Bearer $token", query.ifBlank { null })
        }
    }

    suspend fun userProfileDetailed(userId: Long? = null): Result<UserProfileDetailedDto> {

        if (token == null) {
            logout()
            return Result.Error(IOException("Not authorized!"))
        }

        return resolveResult {
            if (userId == null)
                apiService.currentUser("Bearer $token")
            else
                apiService.userProfileById(userId, "Bearer $token")
        }
    }

    suspend fun login(email: String, password: String): Result<UserTokenFormResult> {

        val form = UserLoginForm(email, password)

        return resolveResult({
            updateToken(it.token)
        }) {
            apiService.login(form)
        }
    }

    suspend fun register(
        email: String,
        username: String,
        photoSrc: String,
        password: String
    ): Result<UserTokenFormResult> {

        val form = UserRegisterForm(
            email, username, photoSrc.ifBlank { "" }, password
        )

        return resolveResult({
            updateToken(it.token)
        }) {
            apiService.register(form)
        }
    }

    fun logout() {
        clearToken()
    }

    val token: String?
        get() = prefs.getString("jwt_token", null)

    private fun clearToken() {
        prefs.edit {
            remove("jwt_token")
        }
    }

    private fun updateToken(token: String) {
        prefs.edit {
            putString("jwt_token", token)
        }
    }

    suspend fun beginTopic(topicId: Long, courseId: Long) : Result<TopicCompendiumDto> {
        return resolveResult {
            apiService.beginTopic("Bearer $token", topicId, courseId)
        }
    }

    suspend fun updateCompendium(compendium: TopicCompendiumDto, courseId: Long) : Result<TopicCompendiumDto> {
        return resolveResult {
            apiService.updateCompendium("Bearer $token", courseId, compendium)
        }
    }

    suspend fun loadCoursePlanById(id: Long): Result<CoursePlanBasicDto> {
        return resolveResult {
            apiService.loadCoursePlanById("Bearer $token", id)
        }
    }

    suspend fun getCardsDailyStats(): Result<FlashCardDayStats> {
        return resolveResult {
            apiService.getCardDailyStats("Bearer $token")
        }
    }

    suspend fun getFlashCardList(): Result<List<FlashCardBasicDto>> {
        return resolveResult {
            apiService.getCurrentCards("Bearer $token")
        }
    }

    suspend fun answerFlashCard(flashCardId: Long, body: FlashCardAnswer)
            : Result<FlashCardBasicDto> {
        return resolveResult {
            apiService.answerFlashCard("Bearer $token", flashCardId, body)
        }
    }

    suspend fun updateConcept(flashCardId: Long, body: TopicCompendiumDto.ConceptBasicDto)
            : Result<FlashCardBasicDto> {
        return resolveResult {
            apiService.updateConcept("Bearer $token", flashCardId, body)
        }
    }

    data class TopicForm(
        var name : MutableState<String> = mutableStateOf(""),
        var desc : MutableState<String> = mutableStateOf("")
    ) {
        constructor (name : String, desc : String) : this(
            mutableStateOf(name),
            mutableStateOf(desc)
        )
    }

    companion object {
        private var retrofit: Retrofit? = null
        private const val BASE_URL = "http://10.0.2.2:8080/api/"
        private const val PROD_URL = "https://cw2025-backend.onrender.com/api/"
        private const val PROD = false
        private const val CONNECT_TIMEOUT = 10L
        private const val READ_TIMEOUT = 60L
        private val client: Retrofit
            get() = retrofit ?: buildClient(if (PROD) PROD_URL else BASE_URL)

        private fun buildClient(path: String): Retrofit {

            val gson = GsonBuilder()
                // Instant deserializer
                .registerTypeAdapter(Instant::class.java, object : JsonDeserializer<Instant> {
                    override fun deserialize(json: JsonElement?, typeOfT: Type?, context: JsonDeserializationContext?): Instant {
                        return Instant.parse(json!!.asString)
                    }
                })
                // Instant serializer
                .registerTypeAdapter(Instant::class.java, object : JsonSerializer<Instant> {
                    override fun serialize(src: Instant?, typeOfSrc: Type?, context: JsonSerializationContext?): JsonElement {
                        return JsonPrimitive(src?.toString())
                    }
                })
                // Duration deserializer
                .registerTypeAdapter(Duration::class.java, object : JsonDeserializer<Duration> {
                    override fun deserialize(json: JsonElement?, typeOfT: Type?, context: JsonDeserializationContext?): Duration {
                        return Duration.parse(json!!.asString)
                    }
                })
                // Duration serializer
                .registerTypeAdapter(Duration::class.java, object : JsonSerializer<Duration> {
                    override fun serialize(src: Duration?, typeOfSrc: Type?, context: JsonSerializationContext?): JsonElement {
                        return JsonPrimitive(src?.toString())
                    }
                })
                .create()

            val client = OkHttpClient.Builder()
                .connectTimeout(CONNECT_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(READ_TIMEOUT, TimeUnit.SECONDS)
                .build()

            retrofit = Retrofit.Builder()
                .baseUrl(path)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()

            return retrofit!!
        }

        private val apiService: ApiService
            get() = client.create<ApiService>(
                ApiService::class.java
            )
    }
}